tutorials/js/content-driven-element.js

```diff
@@ -0,0 +1,330 @@
+            }
+            // label metrics
+            let $label; {
+                let w, h;
+                if ('label' in cache) {
+                    w = cache.label.width;
+                    h = cache.label.height;
+                } else {
+                    const {
+                        width,
+                        height
+                    } = measureText(svg, attributes.label, labelAttributes);
+                    w = width;
+                    h = height;
+                    cache.label = {
+                        width,
+                        height
+                    };
+                }
+                width = Math.max(width, spacing + w + spacing);
+                height += h;
+                if (!h) {
+                    // no text
+                    height -= spacing;
+                }
+                $label = {
+                    x,
+                    y,
+                    width: w,
+                    height: h
+                };
+            }
+            // root metrics
+            return {
+                x: 0,
+                y: 0,
+                width,
+                height,
+                $image,
+                $label
+            };
+        }
+    }
+      
+    const ElementView = joint.dia.ElementView;
+    
+    const ShapeView = ElementView.extend({
+    
+        presentationAttributes: ElementView.addPresentationAttributes({
+            // attributes that changes the position and size of the DOM elements
+            label: [ElementView.Flags.UPDATE],
+            image: [ElementView.Flags.UPDATE],
+            // attributes that do not affect the size
+            outlineColor: ['@color'],
+            fillColor: ['@color'],
+        }),
+        
+        confirmUpdate: function(...args) {
+            let flags = ElementView.prototype.confirmUpdate.call(this, ...args);
+            if (this.hasFlag(flags, '@color')) {
+                // if only a color is changed, no need to resize the DOM elements
+                this.updateColors();
+                flags = this.removeFlag(flags, '@color');
+            }
+            // must return 0
+            return flags;
+        },
+        
+        /* Runs only once while initializing */
+        render: function() {
+            const {
+                vel,
+                model
+            } = this;
+            const body = this.vBody = V('rect').addClass('body');
+            const label = this.vLabel = V('text').addClass('label').attr(model.labelAttributes);
+            this.vImage = V('image').addClass('image').attr(model.imageAttributes);
+            vel.empty().append([
+                body,
+                label
+            ]);
+            this.update();
+            this.updateColors();
+            this.translate(); // default element translate method
+        },
+        
+        update: function() {
+            const layout = this.model.layout();
+            this.updateBody();
```

Comment:
@kumilingus I removed `this.updateBody(layout)` (calling with `layout`) here, not sure if you want it there